File: problems/3904/solution.cpp
#include <vector>
using namespace std;

class Solution {
 public:
  int firstStableIndex(vector<int>& nums, int k) {
    vector<int> preMax(nums.size(), 0);
    vector<int> suMin(nums.size(), 0);
    int M = nums[0];
    int m = nums[nums.size() - 1];
    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] >= M) {
        M = nums[i];
      }
      preMax[i] = M;
      if (nums[nums.size() - 1 - i] <= m) {
        m = nums[nums.size() - 1 - i];
      }
      suMin[nums.size() - 1 - i] = m;
    }
    for (int i = 0; i < nums.size(); i++) {
      if (preMax[i] - suMin[i] <= k) {
        return i;
      }
    }
    return -1;
  }
};

File: problems/3904/solution.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        preMax = [0] * len(nums)
        suMin = [0] * len(nums)
        M = nums[0]
        m = nums[-1]
        for i in range(len(nums)):
            M = max(M, nums[i])
            preMax[i] = M
            m = min(m, nums[len(nums) - 1 - i])
            suMin[len(nums) - 1 - i] = m
        for i in range(len(nums)):
            if preMax[i] - suMin[i] <= k:
                return i
        return -1
